bootstrap/v2/pkg/apis/apps/kfdef/v1alpha1/application_types.go

```diff
@@ -179,3 +265,172 @@ func GetDefaultRegistry() *RegistryConfig {
+// interface. Returns an error if the plugin isn't defined or if there is a problem
+// unmarshaling it.
+func (d *KfDef) GetPluginSpec(pluginName string, s interface{}) error {
+	for _, p := range d.Spec.Plugins {
+		if p.Name != pluginName {
+			continue
+		}
+
+		// To deserialize it to a specific type we need to first serialize it to bytes
+		// and then unserialize it.
+		specBytes, err := yaml.Marshal(p.Spec)
+
+		if err != nil {
+			log.Errorf("Could not marshal plugin %v args; error %v", pluginName, err)
+			return err
+		}
+
+		err = yaml.Unmarshal(specBytes, s)
+
+		if err != nil {
+			log.Errorf("Could not unmarshal plugin %v to the provided type; error %v", pluginName, err)
+		}
+		return nil
+	}
+
+	return NewPluginNotFound(pluginName)
+}
+
+// GetPluginSpec will try to unmarshal the spec for the specified plugin to the supplied
+// interface. Returns an error if the plugin isn't defined or if there is a problem
+// unmarshaling it
+//
+// TODO(jlewi): The reason this function exists is because for types like Gcp in gcp.go
+// we embed KfDef into the Gcp struct so its not actually a type KfDef. In the future
+// we will probably refactor KfApp into an appropriate plugin in type an stop embedding
+// KfDef in it.
+func (s *KfDefSpec) GetPluginSpec(pluginName string, pluginSpec interface{}) error {
+	d := &KfDef{
+		Spec: *s,
+	}
+	return d.GetPluginSpec(pluginName, pluginSpec)
+}
+
+// SetPluginSpec sets the requested parameter. The plugin is added if it doesn't already exist.
+func (d *KfDef) SetPluginSpec(pluginName string, spec interface{}) error {
+	// Convert spec to RawExtension
+
+	r := &runtime.RawExtension{}
+
+	// To deserialize it to a specific type we need to first serialize it to bytes
+	// and then unserialize it.
+	specBytes, err := yaml.Marshal(spec)
+
+	if err != nil {
+		log.Errorf("Could not marshal spec; error %v", err)
+		return err
+	}
+
+	err = yaml.Unmarshal(specBytes, r)
+
+	if err != nil {
+		log.Errorf("Could not unmarshal plugin to RawExtension; error %v", err)
+	}
+
+	index := -1
+
+	for i, p := range d.Spec.Plugins {
+		if p.Name == pluginName {
+			index = i
+			break
+		}
+	}
+
+	if index == -1 {
+		// Plugin in doesn't exist so add it
+		log.Infof("Adding plugin %v", pluginName)
+
+		d.Spec.Plugins = append(d.Spec.Plugins, Plugin{
+			Name: pluginName,
+		})
+
+		index = len(d.Spec.Plugins) - 1
+	}
+
+	d.Spec.Plugins[index].Spec = r
+
+	return nil
+}
+
+// WriteToFile write the KfDef to a file.
+// WriteToFile will strip out any literal secrets before writing it
+func (d *KfDef) WriteToFile(path string) error {
+
+	stripped := *d
+
+	secrets := make([]Secret, 0)
+
+	for _, s := range stripped.Spec.Secrets {
+		if s.SecretSource.LiteralSource != nil {
+			log.Warnf("Stripping literal secret %v from KfDef before serializing it", s.Name)
+			continue
+		}
+		secrets = append(secrets, s)
+	}
+
+	stripped.Spec.Secrets = secrets
```

Comment:
Here the KfDef object was changed. Is it expected?

Reply:
We're making a copy of the spec and then modifying the copy to strip out secrets before writing it to disk.